boa/src/syntax/lexer/number.rs

```diff
@@ -0,0 +1,349 @@
+
+        let exp_str = &mut String::new();
+
+        // The non-digit character could be:
+        // 'n' To indicate a BigIntLiteralSuffix.
+        // '.' To indicate a decimal seperator.
+        // 'e' | 'E' To indicate an ExponentPart.
+        match cursor.peek() {
+            Some(Ok('n')) => {
+                // DecimalBigIntegerLiteral
+                // Lexing finished.
+
+                // Consume the n
+                cursor.next();
+
+                kind = kind.to_bigint();
+            }
+            Some(Ok('.')) => {
+                if kind.base() == 10 {
+                    // Only base 10 numbers can have a decimal seperator.
+                    // Number literal lexing finished if a . is found for a number in a different base.
+
+                    cursor.next();
+                    buf.push('.'); // Consume the .
+                    kind = NumericKind::Rational;
+
+                    // Consume digits until a non-digit character is encountered or all the characters are consumed.
+                    cursor.take_until_pred(&mut buf, &|c: char| c.is_digit(kind.base()))?;
+
+                    // The non-digit character at this point must be an 'e' or 'E' to indicate an Exponent Part.
+                    // Another '.' or 'n' is not allowed.
+                    match cursor.peek() {
+                        Some(Ok('e')) | Some(Ok('E')) => {
+                            // Consume the ExponentIndicator.
+                            cursor.next();
+
+                            buf.push('E');
+
+                            take_signed_integer(&mut buf, cursor, &kind)?;
+                        }
+                        Some(Err(e)) => {
+                            return Err(e.into());
+                        }
+                        Some(Ok(_)) | None => {
+                            // Finished lexing.
+                        }
+                    }
+                }
+            }
+            Some(Ok('e')) | Some(Ok('E')) => {
+                cursor.next(); // Consume the ExponentIndicator.
+                take_signed_integer(exp_str, cursor, &kind)?;
+            }
+            Some(Err(e)) => {
+                return Err(e.into());
+            }
+
+            Some(Ok(_)) | None => {
+                // Indicates lexing finished.
+            }
+        }
+
+        check_after_numeric_literal(cursor)?;
+
+        let num = match kind {
+            NumericKind::BigInt(base) => {
+                Numeric::BigInt(
+                    BigInt::from_string_radix(&buf, base as u32).expect("Could not convert to BigInt")
+                    )
+            }
+            NumericKind::Rational /* base: 10 */ => {
+                match f64::from_str(&buf) {
+                    Ok(val) => Numeric::Rational(val),
+                    Err(e) => return Err(Error::syntax(format!("Unable to parse rational number, reason: {}", e.to_string())))
+                }
+            }
```

Comment:
Here we should use `expect` becasue we want it to crash if it fails to lex, we already checked if it was valid so this cannot fail, if it does that means our floating point lexing logic is faulty. this can become a very hard bug.

Reply:
Ah ok, this makes sense - will make changes